### ISSUE TITLE
[DX11] passing float3x3 constant parameter rearranges values

### ISSUE THREAD

**Issue by marek-g** [gfx-rs/gfx#1161]
Matrix values are rearranged when I pass float3x3 (`[[f32; 3]; 3]`) constant parameter to a vertex shader.

Passing following (identity) matrix
`[[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]`
gives me
`[[1.0, 1.0, 1.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]]`
on the shader side.

As a workaround I can pass float4x4 matrix, which works fine. Ideally, I would like to pass float3x2 or float2x3 (I'm trying to do 2d rotation&translation) but it is not supported now, I think (because of missing gfx::shade::ToUniform implementation?).

_gfx 0.14.0, gfx_core 0.6.0, gfx_device_dx11 0.5.0, gfx_window_dxgi 0.6.0_

**Comment by kvark** [comment]
Thanks for the report!
If you want to assist the investigation, please provide an `apitrace` of such behavior.

**Comment by DaseinPhaos** [comment]
I don't know if you've figured this out or not, but perhaps 

- it has something to do with the memory layout of Rust. Basically when dealing with FFI one should always add the `#[repr(C)]` attribute.

- also if I'm remembering D3D11-stuff correctly, value-types defined in a `cbuffer` in HLSL are always packed into some 16-bytes "vector"s. Thus when updating a 
```
[[1.0, 0.0, 0.0],
  [0.0, 1.0, 0.0],
  [0.0, 0.0, 1.0]]
``` 
into the `cbuffer` (assuming `#[repr(C)]`), which might correspond to a `float3x3 matrix;` (or something like that) in the shader, from the perspective of this `matrix` its interpretation goes like 
```
[[1.0, 0.0, 0.0, /*last value padding*/0.0], // => (1, 0, 0)
 [1.0, 0.0, 0.0, 0.0/*last value padding*/], // => another (1, 0, 0)
 [1.0, ?, ?, ? /*last value padding*/]] // `?`s represent undefined values which in your case happens to also be zero. Thus yet another (1, 0, 0)
```
Basically when working with `cbuffer` you should always consider align your variables such that nothing stretches through a 16-bytes boundary while not being 16-bytes aligned. A more accurate and thorough reference: http://gamedev.stackexchange.com/questions/55317/constant-buffer-content-packing
The reason that you're getting one `(1, 1, 1)` instead of three `(1, 0, 0)` that I'm assuming here is yet another issue, relating to the default column-major assumption of HLSL when consuming matrix.

Personally I think the second possibility is more likely to be the actual reason for this issue, which has little thing to do with gfx or Rust in general.

**Comment by marek-g** [comment]
Thank you Dasein for your answer.  Yes, I was suspecting the problem with data align in the first place.

The column-major vs row-major swap is not a problem for me of course and I see it's not related with gfx or Rust (and it could be my mistake when visualizing data as RGB pixels in shader - I didn't know about the apitrace tool).

There is still a problem that gfx either:
- do not fill the padding values for me when I'm passing [[f32; 3]; 3]
- or do not allow me to use [[f32; 4]; 4] instead [[f32; 3]; 3] with float3x3 type to allow me to put padding values myself (I'm not sure now but I think I tried it and it won't compile)

I was thinking about using 3x2 matrix to have better performance (less data to send?) as this is all I need for 2d operations. But if gfx needs to add padding values to it I can imagine it will be even slower (need to copy the whole array?). So I'm good with using 4x4 floats. 4x2 probably would be the best here (maybe I'll figure out how to add missing gfx::shade::ToUniform variants). It's definitely a low priority. Even less than I thought because it may not help to improve performance. Thanks!
